src/tests/tpch/sb_tpch.c

```diff
@@ -0,0 +1,404 @@
+    char *exec_path = NULL;
+    char **args = NULL;
+    int status = 0;
+    pid_t pid = fork();
+
+    if (pid == -1) {
+        return 1;
+    }
+    if (pid == 0) {
+        path = add_path_to_root("src/tests/tpch/scripts");
+        if (path == NULL)
+            return 1;
+        chdir(path);
+
+        exec_path = malloc(sizeof(char) * (strlen(path) + strlen("/tpch_init.sh")));
+        if (exec_path == NULL)
+            return 1;
+        strcat(exec_path, path);
+        strcat(exec_path, "/tpch_init.sh");
+
+        args = get_script_arguments(exec_path);
+        if (args == NULL)
+            return 1;
+        int exit_code = execve(exec_path, args, sb_globals.env);
+        if (exit_code == -1)
+            printf("Something went wrong: %s\n", strerror(errno));
+        exit(exit_code);
+    }
+    waitpid(pid, &status, 0);
+
+    free(path);
+    free(exec_path);
+
+    if (WIFEXITED(status))
+        return WEXITSTATUS(status);
+    return 0;
+}
+
+static char *get_sql_file_content(unsigned int id)
+{
+    FILE *f = NULL;
+    long file_size = 0;
+    char *content = NULL;
+    char *file_path = malloc(sizeof(char) * (strlen(tpch.query_path) + strlen("/00.sql")));
+
+    if (file_path == NULL)
+        return NULL;
+    if (id < 10)
+        sprintf(file_path, "%s/0%d.sql", tpch.query_path, id);
+    else
+        sprintf(file_path, "%s/%d.sql", tpch.query_path, id);
+
+    f = fopen(file_path, "rb");
+    if (f == NULL) {
+        free(file_path);
+        return NULL;
+    }
+    fseek(f, 0L, SEEK_END);
+    file_size = ftell(f);
+    rewind(f);
+
+    content = malloc(sizeof(char) * file_size);
+    if (content == NULL) {
+        fclose(f);
+        free(file_path);
+        return NULL;
+    }
+
+    fread(content, file_size, 1, f);
+    fclose(f);
+    f = NULL;
+
+    return content;
+}
+
+int register_test_tpch(sb_list_t * tests)
+{
+    SB_LIST_ADD_TAIL(&tpch_test.listitem, tests);
+
+    return 0;
+}
+
+int tpch_prepare(void)
+{
+    if (get_tpch_args() > 0) {
+        return 1;
+    }
+    execute_init_script();
+
+    return 0;
+}
+
+int tpch_init(void)
+{
+    if (get_tpch_args() > 0)
+        return 1;
+
+    tpch.db_driver = db_create(NULL);
+    if (tpch.db_driver == NULL)
+        return 1;
+
+    tpch.query_path = add_path_to_root("src/tests/tpch/scripts/queries");
+    if (tpch.query_path == NULL)
+        return 1;
+
+    return 0;
+}
+
+
+sb_event_t tpch_next_event(int thread_id)
+{
+    sb_event_t req;
+
+    req.type = SB_REQ_TYPE_SQL;
+
+    return req;
+}
+
+int tpch_execute_event(sb_event_t *r, int thread_id)
+{
+    static unsigned int event_count = 1;
+    db_conn_t *conn = NULL;
+    char *query = NULL;
+    size_t len = 0;
+    unsigned int script_id = event_count % 22 == 0 ? 22 : event_count % 22;
+    (void)r; /* unused */
+
+    event_count++;
+
+    conn = db_connection_create(tpch.db_driver);
+    if (conn == NULL) {
+        return 1;
+    }
+
+    query = get_sql_file_content(script_id);
```

Comment:
This is not being freed anywhere. Are we leaking memory>

Reply:
In fact, it seems very wasteful to keep loading the SQL queries from disk on every single `execute_event`. Maybe we could load them all once at startup and keep them in memory in an array, so running queries is as simple as `DISK_QUERIES[script_id]`.

Reply:
Yes, that is a good idea. I followed the suggestion: now the `tpch_init` function loads all queries into memory